internal/controllers/lvmcluster/status.go

```diff
@@ -210,3 +183,113 @@ func translateReasonToState(reason string, currentState lvmv1alpha1.LVMStateType
+	if err == nil {
+		setVolumeGroupsReadyConditionTrue(instance)
+		return
+	} else {
+		logger.Error(err, "failed to validate device class setup")
+	}
+
+	degraded := false
+	for _, nodeItem := range vgNodeStatusList.Items {
+		for _, vgStatus := range nodeItem.Spec.LVMVGStatus {
+			switch vgStatus.Status {
+			case lvmv1alpha1.VGStatusFailed:
+				setVolumeGroupsReadyConditionFailed(instance)
+				return
+			case lvmv1alpha1.VGStatusDegraded:
+				degraded = true
+			}
+		}
+	}
+
+	if degraded {
+		setVolumeGroupsReadyConditionDegraded(instance)
+	}
+}
+
+func validateDeviceClassSetup(cluster *lvmv1alpha1.LVMCluster, nodes *corev1.NodeList, nodeStatusList *lvmv1alpha1.LVMVolumeGroupNodeStatusList) error {
+	for _, deviceClass := range cluster.Spec.Storage.DeviceClasses {
+		validNodeExists := false
+		for _, node := range nodes.Items {
+			valid, err := isNodeValid(&node, cluster, &deviceClass)
+			if err != nil {
+				return err
+			}
+			if !valid {
+				continue
+			}
+
+			// If we reach this point, the node is valid, so check for the NodeStatus
+			validNodeExists = true
+			var relatedNodeStatus *lvmv1alpha1.LVMVolumeGroupNodeStatus
+			for _, nodeStatus := range nodeStatusList.Items {
+				if node.Name == nodeStatus.Name {
+					relatedNodeStatus = &nodeStatus
+					break
+				}
+			}
+
+			// If no node status is found, return an error, we assume it should have been
+			// created by the LVMCluster controller
+			if relatedNodeStatus == nil {
+				return fmt.Errorf("no node status found for node %s,"+
+					"that is part of the expected nodes for device class %s",
+					node.Name, deviceClass.Name)
+			}
+
+			// Check if the VGStatus for the device class is present in the NodeStatus
+			var relatedVGStatus *lvmv1alpha1.VGStatus
+			for _, vgStatus := range relatedNodeStatus.Spec.LVMVGStatus {
+				if vgStatus.Name == deviceClass.Name {
+					relatedVGStatus = &vgStatus
+					break
+				}
+			}
+
+			// If no VGStatus is found, return an error, we assume it should have been
+			// created by the vgmanager
+			if relatedVGStatus == nil {
+				return fmt.Errorf("no VGStatus found for VG %s on node %s,"+
+					"that is part of the expected nodes for device class %s",
+					deviceClass.Name, node.Name, deviceClass.Name)
+			}
+
+			if relatedVGStatus.Status != lvmv1alpha1.VGStatusReady {
+				return fmt.Errorf("VG %s on node %s is not in ready state (%s),"+
+					"that is part of the expected nodes for device class %s",
+					deviceClass.Name, relatedVGStatus.Status, node.Name, deviceClass.Name)
+			}
+		}
+		if !validNodeExists {
+			return fmt.Errorf("no valid node found for device class %s",
+				deviceClass.Name)
+		}
+	}
+
+	return nil
+}
+
+func isNodeValid(node *corev1.Node, cluster *lvmv1alpha1.LVMCluster, deviceClass *lvmv1alpha1.DeviceClass) (bool, error) {
+	// Check if node tolerates all taints
+	if !selector.ToleratesAllTaints(node.Spec.Taints, cluster.Spec.Tolerations) {
```

Comment:
what happens if taints or tolrations are nil?

Reply:
I actually have a test case for this: https://github.com/openshift/lvm-operator/pull/636/files#diff-c50dd555085f94f16659d15a7a8e47f8b59c9ce6531ca5a858bd1c7309662df5R51